crates/wasi-common/src/wasi.rs

```diff
@@ -2,118 +2,474 @@
-    | __WASI_RIGHTS_PATH_FILESTAT_SET_TIMES
-    | __WASI_RIGHTS_FD_FILESTAT_GET
-    | __WASI_RIGHTS_FD_FILESTAT_SET_TIMES
-    | __WASI_RIGHTS_PATH_SYMLINK
-    | __WASI_RIGHTS_PATH_UNLINK_FILE
-    | __WASI_RIGHTS_PATH_REMOVE_DIRECTORY
-    | __WASI_RIGHTS_POLL_FD_READWRITE;
-pub(crate) const RIGHTS_DIRECTORY_INHERITING: __wasi_rights_t =
-    RIGHTS_DIRECTORY_BASE | RIGHTS_REGULAR_FILE_BASE;
-
-// Operations that apply to regular files.
-pub(crate) const RIGHTS_REGULAR_FILE_BASE: __wasi_rights_t = __WASI_RIGHTS_FD_DATASYNC
-    | __WASI_RIGHTS_FD_READ
-    | __WASI_RIGHTS_FD_SEEK
-    | __WASI_RIGHTS_FD_FDSTAT_SET_FLAGS
-    | __WASI_RIGHTS_FD_SYNC
-    | __WASI_RIGHTS_FD_TELL
-    | __WASI_RIGHTS_FD_WRITE
-    | __WASI_RIGHTS_FD_ADVISE
-    | __WASI_RIGHTS_FD_ALLOCATE
-    | __WASI_RIGHTS_FD_FILESTAT_GET
-    | __WASI_RIGHTS_FD_FILESTAT_SET_SIZE
-    | __WASI_RIGHTS_FD_FILESTAT_SET_TIMES
-    | __WASI_RIGHTS_POLL_FD_READWRITE;
-pub(crate) const RIGHTS_REGULAR_FILE_INHERITING: __wasi_rights_t = 0;
-
-// Operations that apply to sockets and socket pairs.
-pub(crate) const RIGHTS_SOCKET_BASE: __wasi_rights_t = __WASI_RIGHTS_FD_READ
-    | __WASI_RIGHTS_FD_FDSTAT_SET_FLAGS
-    | __WASI_RIGHTS_FD_WRITE
-    | __WASI_RIGHTS_FD_FILESTAT_GET
-    | __WASI_RIGHTS_POLL_FD_READWRITE
-    | __WASI_RIGHTS_SOCK_SHUTDOWN;
-pub(crate) const RIGHTS_SOCKET_INHERITING: __wasi_rights_t = RIGHTS_ALL;
-
-// Operations that apply to TTYs.
-pub(crate) const RIGHTS_TTY_BASE: __wasi_rights_t = __WASI_RIGHTS_FD_READ
-    | __WASI_RIGHTS_FD_FDSTAT_SET_FLAGS
-    | __WASI_RIGHTS_FD_WRITE
-    | __WASI_RIGHTS_FD_FILESTAT_GET
-    | __WASI_RIGHTS_POLL_FD_READWRITE;
-#[allow(unused)]
-pub(crate) const RIGHTS_TTY_INHERITING: __wasi_rights_t = 0;
-
-pub fn whence_to_str(whence: __wasi_whence_t) -> &'static str {
-    match whence {
-        __WASI_WHENCE_CUR => "__WASI_WHENCE_CUR",
-        __WASI_WHENCE_END => "__WASI_WHENCE_END",
-        __WASI_WHENCE_SET => "__WASI_WHENCE_SET",
-        other => panic!("Undefined whence value {:?}", other),
+use crate::WasiCtx;
+
+wiggle::from_witx!({
+    witx: ["crates/wasi-common/wig/WASI/phases/snapshot/witx/wasi_snapshot_preview1.witx"],
+    ctx: WasiCtx,
+});
+
+use crate::hostcalls_impl;
+use wiggle_runtime::{GuestPtr, GuestPtrMut, GuestString};
+
+impl wiggle_runtime::GuestErrorType for types::Errno {
+    type Context = WasiCtx;
+
+    fn success() -> types::Errno {
+        types::Errno::Success
+    }
+
+    fn from_error(e: wiggle_runtime::GuestError, ctx: &mut WasiCtx) -> types::Errno {
+        eprintln!("Guest Error: {:?}", e);
+        types::Errno::Inval
     }
 }
 
-pub const __WASI_DIRCOOKIE_START: __wasi_dircookie_t = 0;
+impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
+    fn args_get(
+        &mut self,
+        _argv: GuestPtrMut<GuestPtrMut<u8>>,
+        _argv_buf: GuestPtrMut<u8>,
+    ) -> std::result::Result<(), types::Errno> {
+        unimplemented!("args_get")
+    }
```

Comment:
So those are the idiomatic interface functions that one has to provide an implementation for when implementing any of the WASI snapshots such as `wasi_snapshot_preview1`. I've left them as stubs for the moment as I wasn't sure whether I want to migrate the contents of `hostcalls_impl` module here, or not. On one hand, it would be much cleaner, on the other, given that the implementation of the hostcalls will most likely be invariant to different WASI snapshot versions, perhaps it'd be best to leave them as they are in `hostcalls_impl` and invoke them in the trait impl. I'm open to suggestions here on how to best handle this. Perhaps @pchickey could explain a bit more the reasoning behind the trait, etc.

Reply:
this trait (one generated per witx module) is the point at which user code hooks into wiggle-generated code. The bodies of these funcs are going to be a subset of the body of each func in hostcalls_impl - e.g. for fd_read, I expect the code that lands in the trait body here to just be https://github.com/bytecodealliance/wasmtime/blob/e8c9cf77a23f0ac665c198f654a048b889d04587/crates/wasi-common/src/hostcalls_impl/fs.rs#L167-L178, the decoding of arguments and encoding of results is taken care of (to the extent possible) by the caller of this trait method.

Reply:
Right! And while I wanted to initially put this logic directly in the trait implementation, I thought that maybe it'll be best to leave them in a separate implementation module that could be shared between snapshots. Thoughts?